### PR TITLE
refactor(meta): proactively GC more stale objects.

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previouly all SST objects are retained for at least `min_sst_retention_time_sec` even they're indeed useless.
This PR proactively GC these objects to reduce storage overhead.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by Li0k — APPROVED** [review]

File: src/meta/src/hummock/manager/gc.rs [hunk 1]
```diff
@@ -501,6 +526,30 @@ impl HummockManager {
...
+            .list_pinned_ssts()
+            .into_iter()
+            .chain(self.all_object_ids_in_time_travel().await?)
+            .collect();
+        let object_ids = object_ids
+            .into_iter()
+            .filter(|s| !pinned_objects.contains(s))
+            .collect_vec();
+        // Retry is not necessary. Full GC will handle these objects eventually.
+        self.delete_objects(object_ids).await?;
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/gc.rs:550]: Wil lit be safer to call `complete_gc_batch` directly here?

File: src/meta/src/hummock/manager/gc.rs [hunk 2]
```diff
@@ -48,6 +48,8 @@ pub(crate) struct GcManager {
     store: ObjectStoreRef,
     path_prefix: String,
     use_new_object_prefix_strategy: bool,
+    /// These objects may still be used by backup or time travel.
+    may_delete_object_ids: parking_lot::Mutex<HashSet<HummockSstableObjectId>>,
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/gc.rs:52]: Given that the minor gc is triggered directly after `create_version_checkpoint` and we can always get the full stale objects from `create_version_checkpoint` directly. Will it be better if we instead let `create_version_checkpoint` returns the stale objects without introducing the mutext protected hash set?

Comment (zwang28) [reply src/meta/src/hummock/manager/gc.rs:52]: may_delete_object_ids may also be modified when proactively GC time travel objects. See latest commits.

Comment (zwang28) [reply src/meta/src/hummock/manager/gc.rs:52]: Will refactor.

**Pull request merged**